wptherml/optdriver.py

```diff
@@ -0,0 +1,1204 @@
+        if "qubo_thickness_optimization" in args:
+            self.qubo_thickness_optimization = True
+            self.combinatorial_optimization = False
+            self.optimization_dictionary = args['optimization_dictionary']
+
+
+    def optimize_qubo(self, fom_function):
+
+        if self.combinatorial_optimization == True:
+            self.optimizer = self._qubo_combinatorial_structure_optimization(self, self.optimization_dictionary, fom_function)
+            self.optimizer.learning_loop(num_to_train=25, num_iterations=10000)
+
+        if self.qubo_thickness_optimization == True:
+            
+            self.optimizer = self._qubo_thickness_and_alloy_optimization(self, optimization_dict=self.optimization_dictionary, fom_func = fom_function)
+            self.optimizer.learning_loop(num_to_train=100, n_epochs = 250, l2_lambda=0.001, num_iterations=400, reduction_factor=2,  l1_lambda =0.000001, K = 25, LR = 0.01)
+
+
+
+    def optimize_bfgs(self):
+        """"
+        Method to wrap the l-bfgs-b optimizer
+        """
+        # initialize x array
+        x_start = self.thickness_array[1:self.number_of_layers-1] * 1e9
+
+        # set bounds
+        bfgs_xmin = self.lower_bound * np.ones(self.number_of_layers-2)
+        bfgs_xmax = self.upper_bound * np.ones(self.number_of_layers-2)
+        # rewrite the bounds in the way required by L-BFGS-B
+        bfgs_bounds = [(low, high) for low, high in zip(bfgs_xmin, bfgs_xmax)]
+
+        fom_start, grad_start = self.compute_fom_and_gradient_from_thickness_array(x_start)
+        print(F' Initial FOM is {fom_start}')
+        print(F' Initial Gradient is {grad_start}')
+
+        ret = minimize(self.compute_fom_and_gradient_from_thickness_array, x_start, method="L-BFGS-B", jac=True, bounds=bfgs_bounds)
+        print(ret.x)
+
+
+    def compute_fom_and_gradient_from_thickness_array(self, x0):
+        """"
+        Method to update the thickness array, the relevant spectra, objective function, and (if supported)
+        the gradient
+        """
+        print(" Computing figure of merit and graident from thickness array ", x0)
+
+        # need to add options to use other figures of merit
+        self.thickness_array[1:self.number_of_layers-1] = x0 * 1e-9
+        self.compute_spectrum()
+        self.compute_selective_mirror_fom()
+        # compute_selective_mirror_fom_gradient calls compute_spectrum_gradient()
+        self.compute_selective_mirror_fom_gradient()
+
+        fom = self.reflection_efficiency
+        grad = self.reflection_efficiency_gradient
+
+        if self.minimization:
+            fom *= 1
+            grad *= 1e-9
+        else:
+            fom *= -1
+            grad *= -1e-9
+
+        print(F' Current FOM is {fom}')
+        print(F' Current Gradient is {grad}')
+        return fom, grad
+
+
+    class _qubo_thickness_and_alloy_optimization:
+
+        def __init__(self, outer_instance, optimization_dict: dict, fom_func):
```

Comment:
If I understand correctly, the way the constructor works, the user will need to define a fom_func outside of calling optdriver.  Ideally, the syntax of computing the fom should be hidden from the user and we should have a few methods in optdriver and the qubo classes like the compute_fom_and_gradient_from_thickness_array method that will be called by the appropriate optimization driver (bfgs, qubo, etc).  Right now I have only one fom in that compute_fom_... method, but I will add more conditionals based on the value of self.objective_function that will compute different foms and gradients (when available).  Can you make this adaptation to the _qubo classes, as well?